### ISSUE TITLE
Access Forbidden responses after a time period in 0.15.6

### ISSUE THREAD

**Issue by narcoticfresh** [pomerium/pomerium#2781]
## What happened?

After updating to 0.15.6, we experience many "access forbidden" responses while trying to access services behind pomerium using Github IDP.

This was not the case with 0.15.5.. i've noticed the changes in #2725, don't know if it could have caused the changes in behavior.

The reason pomerium throws access forbidden is the empty groups list at the time of the request. It seems pomerium just "forgets" the group list - then we need to wait for a couple of minutes (i guess sync interval?) and voilà - it works again. But why does pomerium just "forget" the group list? How can that happen?

My assumption is, that the group list (once initialized) stays persisted and gets refreshed every <sync-interval> minutes, thus this state should never happen.

I thought initially that databroker got reset or something (which it turns out was not the case). Thus i configured a redis instance and made sure it got data in it. Also that didn't change anything.

## What did you expect to happen?

That we don't get Forbidden responses when a valid user tries to access a resource. The group list of an user stays as it is and gets refreshed, but not reset (once initialized).

## How'd it happen?

1. Deploy pomerium 0.15.6 with the provided config
2. Access a service, with for initial synchronization
3. Use the service
4. Wait a day
5. Come back, try to access the service again
6. Read "access forbidden" 
7. Wait some minutes, try again
8. Now, it works.. you can use the service

## What's your environment like?

- Pomerium version: 0.15.6
- k3s cluster of 8 nodes, running on alpine linux

## What's your config.yaml?

```config.yaml
# Main configuration flags : https://www.pomerium.io/docs/reference/reference/
insecure_server: true
grpc_insecure: true
address: ":80"
grpc_address: ":80"

pomerium_debug: true

authenticate_service_url: https://**REDACTED**
authorize_service_url: http://pomerium-authorize-service.pomerium.svc.cluster.local

databroker_service_url: http://pomerium-databroker-service.pomerium.svc.cluster.local
# databroker storage
databroker_storage_type: redis
databroker_storage_connection_string: "redis://pomerium-redis:6379/0"

#override_certificate_name: "*.localhost.pomerium.io"
idp_provider: "github"
idp_client_id: "**REDACTED**"
idp_client_secret: "**REDACTED**"
idp_refresh_directory_timeout: 9m
idp_refresh_directory_interval: 10m

idp_service_account: "**REDACTED**"

jwt_claims_headers: "email,groups,user"

policy:
  # general jenkins
  - from: https://**REDACTED**
    to: http://**REDACTED**.platform.svc.cluster.local
    pass_identity_headers: true
    jwt_claims_headers: email, groups, user
    allowed_groups:
      - **REDACTED** (slug1)
      - **REDACTED** (slug2)
  
```

## What did you see in the logs?

I activated debug logging a while ago, this is a typical part of the log with redacted usernames

I hoped by activating debug i could see more information regarding the "failed to update user info" ERRs, but that was not the case..

```logs
9:11AM ERR failed to update user info, deleting session error="github: could not retrieve user info Unauthorized" session_id=309e8d2d-7c57-42b5-b721-cbabb1e68cea user_id=**REDACTED**
9:11AM INF put id=309e8d2d-7c57-42b5-b721-cbabb1e68cea type=type.googleapis.com/session.Session
9:11AM ERR failed to update user info, deleting session error="github: could not retrieve user info Unauthorized" session_id=686a7d01-1f78-4496-976b-ac95e118ccf7 user_id=**REDACTED**
9:11AM INF put id=686a7d01-1f78-4496-976b-ac95e118ccf7 type=type.googleapis.com/session.Session
9:11AM INF put id=**REDACTED** type=type.googleapis.com/user.User
9:11AM ERR failed to update user info, deleting session error="github: could not retrieve user info Unauthorized" session_id=b38f08a9-e4d3-4b81-8b3a-907d5ccc6611 user_id=**REDACTED**
9:11AM INF put id=b38f08a9-e4d3-4b81-8b3a-907d5ccc6611 type=type.googleapis.com/session.Session
9:11AM ERR failed to update user info, deleting session error="github: could not retrieve user info Unauthorized" session_id=c957aadf-9b17-4b94-a06e-f6c23ed2711a user_id=**REDACTED**
9:11AM INF put id=c957aadf-9b17-4b94-a06e-f6c23ed2711a type=type.googleapis.com/session.Session
9:13AM INF refreshing user service=identity_manager user_id=**REDACTED**
9:13AM INF put id=**REDACTED** type=type.googleapis.com/user.User
9:13AM ERR failed to update user info, deleting session error="github: could not retrieve user info Unauthorized" session_id=a61bcbee-ea79-4efd-9b84-eb36cace019b user_id=**REDACTED**
9:13AM INF put id=a61bcbee-ea79-4efd-9b84-eb36cace019b type=type.googleapis.com/session.Session
9:13AM ERR failed to update user info, deleting session error="github: could not retrieve user info Unauthorized" session_id=add4b199-b6a5-431e-8432-b91b5632cd87 user_id=**REDACTED**
9:13AM INF put id=add4b199-b6a5-431e-8432-b91b5632cd87 type=type.googleapis.com/session.Session
9:13AM INF refreshing directory users service=identity_manager
9:15AM INF refreshing user service=identity_manager user_id=**REDACTED**
9:15AM INF put id=**REDACTED** type=type.googleapis.com/user.User
9:17AM INF refreshing user service=identity_manager user_id=**REDACTED**
9:17AM INF refreshing user service=identity_manager user_id=**REDACTED**
9:18AM INF refreshing user service=identity_manager user_id=**REDACTED**
9:18AM INF put id=**REDACTED** type=type.googleapis.com/user.User
9:19AM INF refreshing user service=identity_manager user_id=**REDACTED**
9:19AM INF put id=**REDACTED** type=type.googleapis.com/user.User
9:19AM INF put id=**REDACTED** type=type.googleapis.com/user.User
9:19AM INF put id=**REDACTED** type=type.googleapis.com/user.User
9:21AM INF refreshing user service=identity_manager user_id=**REDACTED**
9:21AM ERR failed to update user info, deleting session error="github: could not retrieve user info Unauthorized" session_id=309e8d2d-7c57-42b5-b721-cbabb1e68cea user_id=**REDACTED**
9:21AM INF put id=309e8d2d-7c57-42b5-b721-cbabb1e68cea type=type.googleapis.com/session.Session
9:21AM ERR failed to update user info, deleting session error="github: could not retrieve user info Unauthorized" session_id=686a7d01-1f78-4496-976b-ac95e118ccf7 user_id=**REDACTED**
9:21AM INF put id=686a7d01-1f78-4496-976b-ac95e118ccf7 type=type.googleapis.com/session.Session
9:21AM INF put id=**REDACTED** type=type.googleapis.com/user.User
9:21AM ERR failed to update user info, deleting session error="github: could not retrieve user info Unauthorized" session_id=b38f08a9-e4d3-4b81-8b3a-907d5ccc6611 user_id=**REDACTED**
9:21AM INF put id=b38f08a9-e4d3-4b81-8b3a-907d5ccc6611 type=type.googleapis.com/session.Session
9:21AM ERR failed to update user info, deleting session error="github: could not retrieve user info Unauthorized" session_id=c957aadf-9b17-4b94-a06e-f6c23ed2711a user_id=**REDACTED**
```

## Additional context

We had no problems like that with 0.15.5..

I though we ran into the github max request limits (which i could not verify); but I've seen that recently the switch to GraphQL happened. So I guess that's not the case? 

Either way, even when running into the request limit, that should not lead to a reset of the groups list IMHO?

And I had to increase the `idp_refresh_directory_timeout` setting as the group fetch took quite long initially. But that change happened already before the update to 0.15.6..

**Comment by travisgroth** [comment]
Hi @narcoticfresh,

Thanks for reporting.  Though I think this needs investigation, we realized that v0.15.6 contains changes which were not intended for a patch release.  We'd recommend upgrading to v0.15.7 in the short term and see if that addresses your issue.  Please let us know either way.

**Comment by narcoticfresh** [comment]
@travisgroth 
thanks for your feedback.. i upgraded to 0.15.7 and will keep an eye on it - will report back here after ~a week to see if we still experience that behavior..

**Comment by narcoticfresh** [comment]
@travisgroth 
great, i didn't see that error anymore.. normally it would have surfaced already.. the update to 0.15.7 solves the issue described here, closing..

thanks!

**Comment by travisgroth** [comment]
Re-opening as this is still likely to be an issue for the next minor release.

**Comment by calebdoxsey** [comment]
I was not able to reproduce this issue. The error logs above imply there was something `Unauthorized`. Perhaps the permissions on the token were wrong? Or the scopes were incorrect?

**Comment by narcoticfresh** [comment]
@calebdoxsey as a side note to this issue.. is there any way to actually log all http communication in the logs? some kind of debug debug mode that also logs all requests and responses that are sent?

if so, it could be documented as part of this issue so it could be analyzed easier next time..

**Comment by narcoticfresh** [comment]
Closed in favor of #2934
